From 5d52f1eb185c3a6c9f4e3f5e7a48ba08f31a78af Mon Sep 17 00:00:00 2001
From: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
Date: Sun, 17 Dec 2023 23:17:20 +0530
Subject: [PATCH] sigma_dut: Add support for SAE-EXT-KEY

Enable SAE-EXT when EHT is enabled for
SAE key management type.

Signed-off-by: Dhanavandhana Kannan <quic_dhanavan@quicinc.com>
---
 ap.c        | 16 ++++++++++++++--
 sigma_dut.h |  1 +
 2 files changed, 15 insertions(+), 2 deletions(-)

--- a/ap.c
+++ b/ap.c
@@ -3120,7 +3120,10 @@ static enum sigma_cmd_result cmd_ap_set_
 			val = get_param(cmd,"KeyMgmtType");
 		if (val) {
 			if (strcasecmp(val, "SAE") == 0) {
-				dut->mbss.conf[non_tx_bss_idx - 1].key_mgmt = AP_WPA2_SAE;
+				if (dut->ap_mode == AP_11be)
+					dut->mbss.conf[non_tx_bss_idx - 1].key_mgmt = AP_SAE_EXT;
+				else
+					dut->mbss.conf[non_tx_bss_idx - 1].key_mgmt = AP_WPA2_SAE;
 			} else {
 				if (check_band(dut, BAND_6G, 0)) {
 					send_resp(dut, conn, SIGMA_INVALID,
@@ -3172,8 +3175,13 @@ static enum sigma_cmd_result cmd_ap_set_
 			val = get_param(cmd, "KeyMgmtType");
 		if (val) {
 			if (strcasecmp(val, "SAE") == 0) {
-				dut->cohosted[dut->cohost_iface].conf
-					[cohosted_bss_index - 1].key_mgmt = AP_WPA2_SAE;
+				if (dut->ap_mode == AP_11be) {
+					dut->cohosted[dut->cohost_iface].conf
+						[cohosted_bss_index - 1].key_mgmt = AP_SAE_EXT;
+				} else {
+					dut->cohosted[dut->cohost_iface].conf
+						[cohosted_bss_index - 1].key_mgmt = AP_WPA2_SAE;
+				}
 			} else if (strcasecmp(val, "WPA2-PSK") == 0 ||
 					(security && strcasecmp(security, "PSK") == 0 &&
 						strcasecmp(val, "WPA2") == 0)) {
@@ -3233,8 +3241,10 @@ static enum sigma_cmd_result cmd_ap_set_
 		} else if (strcasecmp(val, "WPA2-SAE") == 0 ||
 			   strcasecmp(val, "SAE") == 0) {
 			dut->ap_key_mgmt = AP_WPA2_SAE;
-			if (dut->ap_mode == AP_11be)
+			if (dut->ap_mode == AP_11be) {
 				dut->ap_cipher = AP_GCMP_256;
+				dut->ap_key_mgmt = AP_SAE_EXT;
+			}
 			else
 				dut->ap_cipher = AP_CCMP;
 			dut->ap_pmf = AP_PMF_REQUIRED;
@@ -4863,6 +4873,7 @@ static int owrt_ap_config_vap(struct sig
 		case AP_WPA_PSK:
 		case AP_WPA2_SAE:
 		case AP_WPA2_PSK_SAE:
+		case AP_SAE_EXT:
 			if (dut->ap_key_mgmt == AP_WPA2_PSK ||
 			    dut->ap_key_mgmt == AP_WPA2_PSK_SAE) {
 				snprintf(buf, sizeof(buf), "psk2");
@@ -5974,6 +5985,7 @@ static int cmd_wcn_ap_config_commit(stru
 	case AP_WPA2_EAP_SHA256:
 	case AP_WPA2_PSK_SHA256:
 	case AP_WPA2_ENT_FT_EAP:
+	case AP_SAE_EXT:
 		/* Not supported */
 		break;
 	}
@@ -7980,6 +7992,7 @@ static int cmd_ath_ap_config_commit(stru
 	case AP_WPA_PSK:
 		case AP_WPA2_SAE:
 		case AP_WPA2_PSK_SAE:
+		case AP_SAE_EXT:
 		if (dut->ap_key_mgmt == AP_WPA2_PSK ||
 		    dut->ap_key_mgmt == AP_WPA2_SAE ||
 		    dut->ap_key_mgmt == AP_WPA2_PSK_SAE)
@@ -8087,6 +8100,7 @@ static int cmd_ath_ap_config_commit(stru
 		case AP_WPA_PSK:
 		case AP_WPA2_SAE:
 		case AP_WPA2_PSK_SAE:
+		case AP_SAE_EXT:
 			if (dut->ap_key_mgmt == AP_WPA2_PSK ||
 			    dut->ap_key_mgmt == AP_WPA2_SAE ||
 			    dut->ap_key_mgmt == AP_WPA2_PSK_SAE)
@@ -9790,11 +9804,13 @@ write_conf:
 	case AP_WPA2_PSK_SAE:
 	case AP_WPA2_PSK_SHA256:
 	case AP_WPA2_FT_PSK:
+	case AP_SAE_EXT:
 		if (dut->ap_key_mgmt == AP_WPA2_PSK ||
 		    dut->ap_key_mgmt == AP_WPA2_SAE ||
 		    dut->ap_key_mgmt == AP_WPA2_PSK_SAE ||
 		    dut->ap_key_mgmt == AP_WPA2_PSK_SHA256 ||
-		    dut->ap_key_mgmt == AP_WPA2_FT_PSK)
+		    dut->ap_key_mgmt == AP_WPA2_FT_PSK ||
+		    dut->ap_key_mgmt == AP_SAE_EXT)
 			fprintf(f, "wpa=2\n");
 		else if (dut->ap_key_mgmt == AP_WPA2_PSK_MIXED)
 			fprintf(f, "wpa=3\n");
@@ -9802,6 +9818,8 @@ write_conf:
 			fprintf(f, "wpa=1\n");
 		if (dut->ap_key_mgmt == AP_WPA2_SAE)
 			key_mgmt = "SAE";
+		else if (dut->ap_key_mgmt == AP_SAE_EXT)
+			key_mgmt = "SAE-EXT-KEY";
 		else if (dut->ap_key_mgmt == AP_WPA2_PSK_SAE)
 			key_mgmt = "WPA-PSK SAE";
 		else if (dut->ap_key_mgmt == AP_WPA2_FT_PSK)
@@ -9820,6 +9838,8 @@ write_conf:
 		case AP_PMF_REQUIRED:
 			if (dut->ap_key_mgmt == AP_WPA2_SAE)
 				key_mgmt = "SAE";
+			else if (dut->ap_key_mgmt == AP_SAE_EXT)
+				key_mgmt = "SAE-EXT-KEY";
 			else if (dut->ap_key_mgmt == AP_WPA2_PSK_SAE)
 				key_mgmt = "WPA-PSK-SHA256 SAE";
 			else
@@ -10669,6 +10689,20 @@ skip_vht_parameters_set:
 				fprintf(f, "sae_pwe=1\n");
 			}
 
+			if (dut->mbss.conf[i].key_mgmt == AP_SAE_EXT) {
+				fprintf(f, "wpa=2\n");
+				fprintf(f, "wpa_key_mgmt=SAE-EXT-KEY\n");
+				if (dut->program == PROGRAM_EHT) {
+					fprintf(f, "wpa_pairwise=GCMP-256\n");
+					fprintf(f, "group_mgmt_cipher=BIP-GMAC-256\n");
+				} else
+					fprintf(f, "wpa_pairwise=CCMP\n");
+				fprintf(f, "ieee80211w=2\n");
+				fprintf(f, "sae_password=%s\n", dut->mbss.conf[i].psk);
+				if (check_band(dut, BAND_6G, conf_counter))
+					fprintf(f, "sae_pwe=1\n");
+			}
+
 			if (dut->mbss.conf[i].key_mgmt == AP_WPA2_PSK) {
 				fprintf(f, "wpa=2\n");
 				fprintf(f, "wpa_key_mgmt=WPA-PSK\n");
@@ -10751,6 +10785,20 @@ skip_vht_parameters_set:
 				fprintf(f, "ieee80211w=2\n");
 				fprintf(f, "sae_password=%s\n",
 					dut->cohosted[dut->cohost_iface].conf[i].psk);
+				fprintf(f, "sae_pwe=1\n");
+			}
+
+			if (dut->cohosted[dut->cohost_iface].conf[i].key_mgmt == AP_SAE_EXT) {
+				fprintf(f, "wpa=2\n");
+				fprintf(f, "wpa_key_mgmt=SAE-EXT-KEY\n");
+				if (dut->program == PROGRAM_EHT) {
+					fprintf(f, "wpa_pairwise=GCMP-256\n");
+					fprintf(f, "group_mgmt_cipher=BIP-GMAC-256\n");
+				} else
+					fprintf(f, "wpa_pairwise=CCMP\n");
+				fprintf(f, "ieee80211w=2\n");
+				fprintf(f, "sae_password=%s\n",
+					dut->cohosted[dut->cohost_iface].conf[i].psk);
 				fprintf(f, "sae_pwe=1\n");
 			}
 
--- a/sigma_dut.h
+++ b/sigma_dut.h
@@ -503,6 +503,7 @@ enum ap_key_mgmt {
 	AP_WPA2_PSK_SHA256,
 	AP_WPA2_ENT_FT_EAP,
 	AP_OSEN,
+	AP_SAE_EXT,
 };
 
 enum eht_ru_alloc {
